File: instagram.ts
import axios from "axios";
import fs from "fs";
import Bluebird from "bluebird";
import dotenv from "dotenv";
dotenv.config();

const getPosts = async ({
  username,
  resultsLimit = 200,
}: {
  username: string;
  resultsLimit?: number;
}) => {
  if (!process.env.APIFY_API_KEY) {
    throw new Error("APIFY_API_KEY is not set in the environment variables");
  }

  const data = {
    username: [username],
    resultsLimit,
  };

  const headers = {
    headers: {
      Authorization: `Bearer ${process.env.APIFY_API_KEY}`,
    },
  };

  try {
    const response = await axios.post(
      `https://api.apify.com/v2/acts/apify~instagram-reel-scraper/run-sync-get-dataset-items`,
      data,
      headers
    );
    if (response.status !== 201) {
      throw new Error("Failed to fetch posts");
    }

    const posts: CleanedInstagramPost[] = response.data.map(
      (post: InstagramPost) => ({
        id: post.id,
        shortCode: post.shortCode,
        timestamp: post.timestamp,
        url: post.url,
        videoUrl: post.videoUrl,
        caption: post.caption,
        hashtags: post.hashtags,
        displayUrl: post.displayUrl,
        likesCount: post.likesCount,
      })
    );
    console.log(posts);

    return posts;
  } catch (error) {
    console.error("Error fetching posts:", error);
    throw error;
  }
};

const downloadIGPosts = async ({
  username,
  posts,
}: {
  username: string;
  posts: CleanedInstagramPost[];
}) => {
  const folderName = `${username}_ig_videos`;

  // Create the directory if it doesn't exist
  if (!fs.existsSync(folderName)) {
    await fs.promises.mkdir(folderName);
  }

  await Bluebird.map(
    posts,
    async (post) => {
      try {
        const response = await axios.get(post.videoUrl, {
          responseType: "arraybuffer",
        });
        const buffer = Buffer.from(response.data, "binary");
        const fileName = `${folderName}/ig_${username}_${post.shortCode}.mp4`;
        console.log(`Saving ${fileName}`);
        await fs.promises.writeFile(fileName, buffer);
      } catch (error) {
        console.error(
          `Error downloading video for post ${post.shortCode}:`,
          error
        );
      }
    },
    { concurrency: 2 }
  );
};

(async () => {
  const username = "oliviacaplandesign";
  console.log("fetching profile data for...", username);

  const resultsLimit = 200;

  try {
    const posts = await getPosts({ username, resultsLimit });
    fs.writeFileSync(
      `ig_${username}_posts.json`,
      JSON.stringify(posts, null, 2)
    );

    console.log(`Fetched ${posts.length} posts for ${username}`);

    console.log("downloading videos...");

    await downloadIGPosts({ username, posts });
  } catch (error) {
    console.error("An error occurred:", error);
  }
})();

interface InstagramPost {
  inputUrl: string;
  id: string;
  type: string;
  shortCode: string;
  caption: string;
  hashtags: string[];
  mentions: string[];
  url: string;
  commentsCount: number;
  firstComment: string;
  latestComments: string[];
  dimensionsHeight: number;
  dimensionsWidth: number;
  displayUrl: string;
  images: string[];
  videoUrl: string;
  alt: string | null;
  likesCount: number;
  videoViewCount: number;
  videoPlayCount: number;
  timestamp: string;
  childPosts: any[]; // Replace `any` with a more specific type if known
  ownerFullName: string;
  ownerUsername: string;
  ownerId: string;
  productType: string;
  videoDuration: number;
  isSponsored: boolean;
  taggedUsers: string[]; // Replace with a more specific type if known
  musicInfo: object; // Replace `object` with a more specific type if known
}

interface CleanedInstagramPost {
  id: string;
  shortCode: string;
  timestamp: string;
  url: string;
  videoUrl: string;
  caption: string;
  hashtags: string[];
  displayUrl: string;
  likesCount: number;
}

File: python-scraper/instagram.py
import os
import json
import requests
from concurrent.futures import ThreadPoolExecutor
from dotenv import load_dotenv

load_dotenv()

def get_posts(username, results_limit=200):
    if not os.getenv("APIFY_API_KEY"):
        raise ValueError("APIFY_API_KEY is not set in the environment variables")

    data = {
        "username": [username],
        "resultsLimit": results_limit,
    }

    headers = {
        "Authorization": f"Bearer {os.getenv('APIFY_API_KEY')}",
    }

    response = requests.post(
        "https://api.apify.com/v2/acts/apify~instagram-reel-scraper/run-sync-get-dataset-items",
        json=data,
        headers=headers
    )

    if response.status_code != 201:
        raise Exception("Failed to fetch posts")

    posts = [
        {
            "id": post["id"],
            "shortCode": post["shortCode"],
            "timestamp": post["timestamp"],
            "url": post["url"],
            "videoUrl": post["videoUrl"],
            "caption": post["caption"],
            "hashtags": post["hashtags"],
            "displayUrl": post["displayUrl"],
            "likesCount": post["likesCount"],
        }
        for post in response.json()
    ]

    print(posts)
    return posts

def download_ig_posts(username, posts):
    folder_name = f"{username}_ig_videos"

    if not os.path.exists(folder_name):
        os.makedirs(folder_name)

    def download_post(post):
        try:
            response = requests.get(post["videoUrl"], stream=True)
            file_name = f"{folder_name}/ig_{username}_{post['shortCode']}.mp4"
            print(f"Saving {file_name}")
            with open(file_name, "wb") as f:
                for chunk in response.iter_content(chunk_size=8192):
                    f.write(chunk)
        except Exception as e:
            print(f"Error downloading video for post {post['shortCode']}: {e}")

    with ThreadPoolExecutor(max_workers=2) as executor:
        executor.map(download_post, posts)

if __name__ == "__main__":
    username = "oliviacaplandesign"
    print(f"fetching profile data for... {username}")

    results_limit = 200

    try:
        posts = get_posts(username, results_limit)
        with open(f"ig_{username}_posts.json", "w") as f:
            json.dump(posts, f, indent=2)

        print(f"Fetched {len(posts)} posts for {username}")
        print("downloading videos...")

        download_ig_posts(username, posts)
    except Exception as e:
        print(f"An error occurred: {e}")